explicit_auto_deref gives bad advice for lazy_static references

### Summary

In 1.65, clippy warns about code that it was fine with in 1.64. The suggested fix does not compile.

When encountering an explicit dereference of a "ref" struct generated by `lazy_static!`, it provides a suggested fix that doesn't work:
```
warning: deref which would be done by auto-deref
  --> src/lib.rs:12:9
   |
12 |     foo(*CONFIG);
   |         ^^^^^^^ help: try this: `CONFIG`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref
   = note: `#[warn(clippy::explicit_auto_deref)]` on by default
```
If I apply that fix then the code no longer compiles:
```
error[E0308]: mismatched types
  --> src/lib.rs:12:9
   |
12 |     foo(CONFIG);
   |     --- ^^^^^^
   |     |   |
   |     |   expected `&str`, found struct `CONFIG`
   |     |   help: consider borrowing here: `&CONFIG`
   |     arguments to this function are incorrect
   |
note: function defined here
  --> src/lib.rs:9:4
   |
9  | fn foo(_: &'static str) {}
   |    ^^^ ---------------
```

The "help" is correct: `&CONFIG` works (with no lints).

### Reproducer

I tried this code:

```rust
fn read_config() -> &'static str {
    "abc"
}

lazy_static::lazy_static! {
    pub static ref CONFIG: &'static str = read_config();
}

fn foo(_: &'static str) {}

pub fn bar() {
    foo(*CONFIG);
}
```

I expected to see this happen:

Either there should be no lint, or the suggested fix should still compile.

Instead, this happened:

The suggest fix does not compile.

### Version

```text
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-unknown-linux-gnu
release: 1.65.0
LLVM version: 15.0.0

It also fails on 1.67.0-nightly (11ebe6512 2022-11-01).
```


### Additional Labels

_No response_

--- a/clippy_lints/src/dereference.rs
+++ b/clippy_lints/src/dereference.rs
@@ -289,23 +289,24 @@ impl<'tcx> LateLintPass<'tcx> for Dereferencing<'tcx> {
                 let (position, adjustments) = walk_parents(cx, &mut self.possible_borrowers, expr, &self.msrv);
                 match kind {
                     RefOp::Deref => {
+                        let sub_ty = typeck.expr_ty(sub_expr);
                         if let Position::FieldAccess {
                             name,
                             of_union: false,
                         } = position
-                            && !ty_contains_field(typeck.expr_ty(sub_expr), name)
+                            && !ty_contains_field(sub_ty, name)
                         {
                             self.state = Some((
                                 State::ExplicitDerefField { name },
                                 StateData { span: expr.span, hir_id: expr.hir_id, position },
                             ));
-                        } else if position.is_deref_stable() {
+                        } else if position.is_deref_stable() && sub_ty.is_ref() {
                             self.state = Some((
                                 State::ExplicitDeref { mutability: None },
                                 StateData { span: expr.span, hir_id: expr.hir_id, position },
                             ));
                         }
-                    }
+                    },
                     RefOp::Method(target_mut)
                         if !is_lint_allowed(cx, EXPLICIT_DEREF_METHODS, expr.hir_id)
                             && position.lint_explicit_deref() =>
@@ -320,7 +321,7 @@ impl<'tcx> LateLintPass<'tcx> for Dereferencing<'tcx> {
                             StateData {
                                 span: expr.span,
                                 hir_id: expr.hir_id,
-                                position
+                                position,
                             },
                         ));
                     },
@@ -394,7 +395,11 @@ impl<'tcx> LateLintPass<'tcx> for Dereferencing<'tcx> {
                                     msg,
                                     snip_expr,
                                 }),
-                                StateData { span: expr.span, hir_id: expr.hir_id, position },
+                                StateData {
+                                    span: expr.span,
+                                    hir_id: expr.hir_id,
+                                    position,
+                                },
                             ));
                         } else if position.is_deref_stable()
                             // Auto-deref doesn't combine with other adjustments
@@ -406,7 +411,7 @@ impl<'tcx> LateLintPass<'tcx> for Dereferencing<'tcx> {
                                 StateData {
                                     span: expr.span,
                                     hir_id: expr.hir_id,
-                                    position
+                                    position,
                                 },
                             ));
                         }

--- a/tests/ui/explicit_auto_deref.fixed
+++ b/tests/ui/explicit_auto_deref.fixed
@@ -277,4 +277,8 @@ fn main() {
         unimplemented!()
     }
     let _: String = takes_assoc(&*String::new());
+
+    // Issue #9901
+    fn takes_ref(_: &i32) {}
+    takes_ref(*Box::new(&0i32));
 }

--- a/tests/ui/explicit_auto_deref.rs
+++ b/tests/ui/explicit_auto_deref.rs
@@ -277,4 +277,8 @@ fn main() {
         unimplemented!()
     }
     let _: String = takes_assoc(&*String::new());
+
+    // Issue #9901
+    fn takes_ref(_: &i32) {}
+    takes_ref(*Box::new(&0i32));
 }

